lib/src/gotrue_client.dart

```diff
@@ -57,89 +66,192 @@ class GoTrueClient {
+      };
+      final fetchOptions = GotrueRequestOptions(headers: _headers, body: body);
+      response = await _fetch.request('$_url/signup', RequestMethodType.post,
+          options: fetchOptions) as Map<String, dynamic>;
+    } else {
+      throw AuthException(
+          'You must provide either an email or phone number and a password');
+    }
+
+    final authResponse = AuthResponse.fromJson(response);
+
+    final session = authResponse.session;
+    if (session != null) {
+      _saveSession(session);
       _notifyAllSubscribers(AuthChangeEvent.signedIn);
     }
 
-    return response;
+    return authResponse;
   }
 
-  /// Signs up a new user using their phone number and a password.
-  ///
-  /// [phone] is the user's phone number WITH international prefix
-  ///
-  /// [password] is the password of the user
-  ///
-  /// [userMetadata] sets [User.userMetadata] without an extra call to [update]
-  Future<GotrueSessionResponse> signUpWithPhone(
-    String phone,
-    String password, {
-    AuthOptions? options,
-    Map<String, dynamic>? userMetadata,
+  /// Log in an existing user with an email and password or phone and password.
+  Future<AuthResponse> signInWithPassword({
+    String? email,
+    String? phone,
+    required String password,
+    String? captchaToken,
   }) async {
     _removeSession();
 
-    final response =
-        await api.signUpWithPhone(phone, password, userMetadata: userMetadata);
-    if (response.session == null) {
-      throw GoTrueException('An error occurred on sign up.');
+    late final Map<String, dynamic> response;
+
+    if (email != null) {
+      response = await _fetch.request(
+        '$_url/token',
+        RequestMethodType.post,
+        options: GotrueRequestOptions(
+          headers: _headers,
+          body: {'email': email, 'password': password},
+          query: {'grant_type': 'password'},
+        ),
+      );
+    } else if (phone != null) {
+      response = await _fetch.request(
+        '$_url/token',
+        RequestMethodType.post,
+        options: GotrueRequestOptions(
+          headers: _headers,
+          body: {'phone': phone, 'password': password},
+          query: {'grant_type': 'password'},
+        ),
+      );
+    } else {
+      throw AuthException(
+        'You must provide either an email, phone number, a third-party provider or OpenID Connect.',
+      );
     }
 
-    _saveSession(response.session!);
-    _notifyAllSubscribers(AuthChangeEvent.signedIn);
+    final authResponse = AuthResponse.fromJson(response);
 
-    return response;
+    if (authResponse.session != null) {
+      _saveSession(authResponse.session!);
+      _notifyAllSubscribers(AuthChangeEvent.signedIn);
+    }
+    return authResponse;
   }
 
-  /// Log in an existing user, or login via a third-party provider.
-  Future<GotrueSessionResponse> signIn({
+  /// Log in an existing user via a third-party provider.
+  Future<OAuthResponse> getOAuthSignInUrl({
```

Comment:
In js client, this method is called `signInWithOAuth`, but I have renamed it `getOAuthSignInUrl`, because all this method is doing is getting the URL. I want to rename the [`signInWithProvider()`](https://github.com/supabase-community/supabase-flutter/blob/main/lib/src/supabase_auth.dart#L303) to `signInWithOAuth()`, which I think is more aligned with the js SDK, but what does everyone think about that?